test/functional/interface_bitcoin_cli.py

```diff
@@ -55,31 +46,50 @@ def run_test(self):
 
-        self.log.info("Make sure that -getinfo with arguments fails")
+        self.log.info("Test -getinfo with arguments fails")
         assert_raises_process_error(1, "-getinfo takes no arguments", self.nodes[0].cli('-getinfo').help)
 
-        self.log.info("Test that -getinfo returns the expected network and blockchain info")
+        self.log.info("Test -getinfo returns expected network and blockchain info")
+        if self.is_wallet_compiled():
+            self.nodes[0].encryptwallet(password)
         cli_get_info = self.nodes[0].cli('-getinfo').send_cli()
         network_info = self.nodes[0].getnetworkinfo()
         blockchain_info = self.nodes[0].getblockchaininfo()
-
         assert_equal(cli_get_info['version'], network_info['version'])
         assert_equal(cli_get_info['blocks'], blockchain_info['blocks'])
+        assert_equal(cli_get_info['headers'], blockchain_info['headers'])
         assert_equal(cli_get_info['timeoffset'], network_info['timeoffset'])
         assert_equal(cli_get_info['connections'], network_info['connections'])
         assert_equal(cli_get_info['proxy'], network_info['networks'][0]['proxy'])
         assert_equal(cli_get_info['difficulty'], blockchain_info['difficulty'])
         assert_equal(cli_get_info['chain'], blockchain_info['chain'])
+
         if self.is_wallet_compiled():
-            self.log.info("Test that -getinfo returns the expected wallet info")
+            self.log.info("Test -getinfo and bitcoin-cli getwalletinfo return expected wallet info")
             assert_equal(cli_get_info['balance'], BALANCE)
             wallet_info = self.nodes[0].getwalletinfo()
             assert_equal(cli_get_info['keypoolsize'], wallet_info['keypoolsize'])
+            assert_equal(cli_get_info['unlocked_until'], wallet_info['unlocked_until'])
             assert_equal(cli_get_info['paytxfee'], wallet_info['paytxfee'])
             assert_equal(cli_get_info['relayfee'], network_info['relayfee'])
-            # unlocked_until is not tested because the wallet is not encrypted
+            assert_equal(self.nodes[0].cli.getwalletinfo(), wallet_info)
         else:
-            self.log.info("*** Wallet not compiled; -getinfo wallet tests skipped")
+            self.log.info("*** Wallet not compiled; cli getwalletinfo and -getinfo wallet tests skipped")
+
+        self.stop_node(0)
+
+        self.log.info("Test -version with node stopped")
+        cli_response = self.nodes[0].cli("-version").send_cli()
+        assert "{} RPC client version".format(self.config['environment']['PACKAGE_NAME']) in cli_response
+
+        self.log.info("Test -rpcwait option waits for RPC connection instead of failing")
+        # Start node without RPC connection.
+        self.nodes[0].start()
+        # Verify failure without -rpcwait.
+        assert_raises_process_error(1, "Could not connect to the server", self.nodes[0].cli('getblockcount').echo)
```

Comment:
Isn't this racy?

Reply:
Will run it in a loop and see if it's flakey.

Reply:
Running `(for i in {1..1000}; do test/functional/interface_bitcoin_cli.py ; done)` to see.

Reply:
We generally only see races in travis or valgrind

Logically, I don't see how the rpc connection is guaranteed to be down when the cli is called here.

Reply:
Makes sense: it ran 1000 times without fail for me in macOS and Debian, but yeah, travis and valgrind. I'll wait to see the travis result for the fun of it and then probably remove that line; we'll still have the assertion with -rpcwait

Reply:
Travis all green. Maybe leave it in and remove if we see issues later?